src/utils/clone.ts

```diff
@@ -0,0 +1,34 @@
+function collectElements(root: HTMLElement): HTMLElement[] {
+    const elements: HTMLElement[] = [];
+    const walker: TreeWalker = document.createTreeWalker(root, NodeFilter.SHOW_ELEMENT, null);
+
+    let element: Node | null = walker.nextNode();
+    while (element) {
+        elements.push(element as HTMLElement);
+        element = walker.nextNode();
+    }
+
+    return elements;
+}
+
+export function cloneShadowRoots(sourceNode: Node, targetNode: Node) {
+
+    const sourceElements = collectElements(sourceNode as HTMLElement);
+    const targetElements = collectElements(targetNode as HTMLElement);
+
+    for (let i = 0; i < sourceElements.length; i++) {
+        const sourceElement = sourceElements[i];
+        const targetElement = targetElements[i];
+
+        const shadowRoot = sourceElement.shadowRoot;
+        if (shadowRoot !== null) {
```

Comment:
Should we guard that `targetElement` is also defined, since it's possible `targetElements` may have a different length than `sourceElements` (this would likely only happen if the user isn't waiting for their DOM to be settled before printing, but if they make a mistake there this will hard crash them. I just added a similar check when iterating over `canvas` elements: https://github.com/MatthewHerbst/react-to-print/blob/8188f8406ca17752a513b5bf46172cf953df3805/src/utils/handlePrintWindowOnLoad.ts#L121-L127)

Reply:
Please review, I added a similar error handling like you suggested.

Reply:
Perfect, thanks!